### ISSUE TITLE
ipv6 forwarding should be global only

### ISSUE THREAD

**Issue by hpresnall** [ifupdown-ng/ifupdown-ng#235]
The current `forward-ipv6 yes` directive enables forwarding only for the interface being managed. However, this doesn't actually configure IPv6 forwarding and incoming packets will be silently dropped after the PREROUTING chain in iptables.

Instead `net.ipv6.conf.all.forwarding` needs to be configured. This is different than ipv4 and could be a cause of confusion. The directive should probably just (re)enable this one value rather than changing anything for the interface.

See https://unix.stackexchange.com/questions/348533/is-net-ipv6-conf-all-forwarding-1-equivalent-to-enabling-forwarding-for-all-indi/348534#348534 and the lined kernel docs.

**Comment by EasyNetDev** [comment]
Hi @hpresnall ,

Normally IPv4 and IPv6 should be activated via sysctl in /etc/sysctl.d/ like this:
```
# cat /etc/sysctl.d/10-ip_forwarding.conf
# Enable IPv4 and IPv6 forwarding.
net.ipv4.ip_forward=1
net.ipv6.conf.all.forwarding=1
net.ipv6.conf.default.forwarding=1
```

All 4 directives `forward-ipv4`, `forward-ipv4-mc`, `forward-ipv6`, `forward-ipv6-mc` are just interface based configurations.

If you check the sources in `ifupdown-ng/executor-scripts/linux/forward`:
```
#!/bin/sh

yesno() {
        case "$1" in
        yes|1)  echo 1 ;;
        *)      echo 0 ;;
        esac
}

[ "$PHASE" != "up" ] && exit 0
[ -z "$VERBOSE" ] || set -x

[ -n "$IF_FORWARD_IPV4" ] && ${MOCK} /bin/sh -c "echo $(yesno $IF_FORWARD_IPV4) > /proc/sys/net/ipv4/conf/$IFACE/forwarding"
[ -n "$IF_FORWARD_IPV6" ] && ${MOCK} /bin/sh -c "echo $(yesno $IF_FORWARD_IPV6) > /proc/sys/net/ipv6/conf/$IFACE/forwarding"

[ -n "$IF_FORWARD_IPV4_MC" ] && ${MOCK} /bin/sh -c "echo $(yesno $IF_FORWARD_IPV4_MC) > /proc/sys/net/ipv4/conf/$IFACE/mc_forwarding"
[ -n "$IF_FORWARD_IPV6_MC" ] && ${MOCK} /bin/sh -c "echo $(yesno $IF_FORWARD_IPV6_MC) > /proc/sys/net/ipv6/conf/$IFACE/mc_forwarding"

exit 0
```

You can see that all directives are just for interfaces.

Even FRRouting is saying that is not "its job" to enable the IP forwarding, although could be a better option to enable it globally.
This option is above `net.ipv4.ip_forward` and `net.ipv6.conf.all.forwarding` / `net.ipv6.conf.default.forwarding`, which can set the interface just as non-routing.

Kind regards,
@EasyNetDev

**Comment by hpresnall** [comment]
I agree ipv6 forwarding should probably be set in /etc/sysctl.d. An interface level setting doesn't make sense.

This fix for this bug would either be to
1) remove support for `forward-ipv6 yes` completely, which may be a breaking change
2) have `forward-ipv6 yes` ignore the interface and always set `net.ipv6.conf.all.forwarding`. If multiple interfaces use this config, re-setting the sysctl variable is redundant but does not have any effect.

**Comment by kaniini** [comment]
The purpose of the config option is to allow it to be set on the interfaces you want to enable forwarding for.  This means that you would use it on input and output interfaces.

Allowing forwarding by default is accomplished through other means and carries the risk of possibly forwarding traffic that shouldn’t be forwarded.

**Comment by EasyNetDev** [comment]
I agree with @kaniini.

The purpose of this config line should be only for the specific interface.
As I mention also, the global forwarding should be done specifically by the admin.

Take an example: you have a Linux box that has multiple purposes. One of them is to have some VPN interfaces. Setting global forwarding could lead to some issues in the system by letting some traffic to be forwarded on some interfaces that you don't want.

In this case you can open a Pandora box without the Admin knowing this.

**Comment by DanielG** [comment]
There may be one possible way to sort of safely enable all.forwarding without causing too much havoc for sysadmins: If we set `default.forwarding=0` and `all.forwarding=1` we should be able to individually opt interfaces into forwarding without affecting all of them.

That being said last time I looked into `default.forwarding` I found the behaviour quite puzzling. I believe it only applies to newly created interfaces and not existing ones, so we'd have to set all exiting interface's forwarding sysctls to =0 at startup.

With enough attention to detail I do think that could be made to work tho.

**Comment by hpresnall** [comment]
The issue is that the per-interface settings _do not_ work for ipv6. all.forwarding must be enabled; setting on a single interface does not do anything.

If you keep the setting in ifupdown then it should enable all (with a logged warning?).

**Comment by EasyNetDev** [comment]
> The issue is that the per-interface settings _do not_ work for ipv6. all.forwarding must be enabled; setting on a single interface does not do anything.
> 
> If you keep the setting in ifupdown then it should enable all (with a logged warning?).

That's an interesting thing. I didn't test it with IPv6 global turned off and per interface.
In this case make sense when an interface is turned on/off to enable global forwarding.

But the problem remains with unconfigured interfaces in ifupdown-ng how to manage with per interface forwarding.

@kaniini any idea?

**Comment by kaniini** [comment]
> The issue is that the per-interface settings do not work for ipv6.

Why not?  The code in the kernel which handles these sysctls appears to be meaningful: turning the sysctl off for an interface disallows packets to be forwarded from it; turning the sysctl on allows packets to be forwarded from it.

What is your definition of "does not work" then?  Is this more of a "you're doing IPv6 *wrong*" take?  Those aren't really convincing.

**Comment by hpresnall** [comment]
Please see the stack overflow link. Based on that, all forwarding is needed.

Setting per interface has no effect. No forwarding happens unless all is enabled. If all is 1, the per interface value can still be 0. Setting per interface to 1 does nothing if all is 0.

Is does not appear to be possible to enable / disable for a single interface.

**Comment by DanielG** [comment]
> Setting per interface has no effect. No forwarding happens unless all is enabled. If all is 1, the per interface value can still be 0. Setting per interface to 1 does nothing if all is 0.

Right and that's what I was trying to point out (poorly). This is where `.default` comes in to make setting per-interface sysctl workable. We turn on `.all=1` and `.default=0`, initialize all interfaces that don't have ipv6-forwarding configured to `.$iface.forwarding=0`. Any new interfaces that show up (think: usb-ethernet adpater) should then be configured to `.forwarding=0` by the kernel  and the user can still enable forwarding for them in the config if desired.

This may be a surprising setup for people that have never seen it, especially since the sysctl behaviour is kind of confusing, but I think we can paper over that with ample documentation.
